### PR TITLE
chore(observability): ensure sinks are checked for component spec compliance

### PR DESCRIPTION
**Pull request by tobz**

### CODE REVIEW

**Review by spencergilbert — APPROVED** [review]
Noting that this PR will also close: https://github.com/vectordotdev/vector/issues/12284

**Review by lukesteensen — APPROVED** [review]

File: src/sinks/datadog_archives.rs [hunk 1]
```diff
@@ -562,41 +570,43 @@ struct DatadogGcsRequestBuilder {
...
-            count: events.len(),
-            byte_size: events.size_of(),
-            finalizers,
-        };
 
-        (metadata, events)
+        (
+            (partition_key, event_count, finalizers, event_byte_size),
+            events,
+        )
```
Comment (bruceg) [inline src/sinks/datadog_archives.rs:589]: This looks like the new metadata has all the same data as the old. What is the motivation for avoiding `GcsMetadata` here?

File: src/sinks/gcp/cloud_storage.rs [hunk 2]
```diff
@@ -220,7 +220,7 @@ struct RequestSettings {
 }
 
 impl RequestBuilder<(String, Vec<Event>)> for RequestSettings {
-    type Metadata = GcsMetadata;
+    type Metadata = (String, usize, EventFinalizers, usize);
```
Comment (bruceg) [inline src/sinks/gcp/cloud_storage.rs:223]: Ditto previous question

File: src/sinks/gcp/stackdriver_logs.rs [hunk 3]
```diff
@@ -36,6 +36,9 @@ enum HealthcheckError {
 #[derive(Deserialize, Serialize, Debug, Clone, Default)]
 #[serde(deny_unknown_fields)]
 pub struct StackdriverConfig {
+    #[serde(skip, default = "default_endpoint")]
+    endpoint: String,
+
```
Comment (bruceg) [inline src/sinks/gcp/stackdriver_logs.rs:41]: I don't see a docs change to add this config item, which should also be noted in the announcement.

File: src/sinks/gcp/stackdriver_metrics.rs [hunk 4]
```diff
@@ -1,3 +1,7 @@
+// TODO: In order to correctly assert component specification compliance, we would have to do some more advanced mocking
+// off the endpoint, which would include also providing a mock OAuth2 endpoint to allow for generating a token from the
+// mocked credentials. Let this TODO serve as a placeholder for doing that in the future.
```
Comment (bruceg) [inline src/sinks/gcp/stackdriver_metrics.rs:3]: Noting here to make sure to add an issue for this TODO once this PR is merged.

File: src/sinks/gcs_common/service.rs [hunk 5]
```diff
@@ -39,28 +41,22 @@ impl GcsService {
...
 #[derive(Clone, Debug)]
 pub struct GcsRequest {
+    pub key: String,
     pub body: Bytes,
     pub settings: GcsRequestSettings,
-    pub metadata: GcsMetadata,
-}
-
-#[derive(Clone, Debug)]
-pub struct GcsMetadata {
```
Comment (bruceg) [inline src/sinks/gcs_common/service.rs:48]: Ah, I see you removed this struct. Is that because you dropped its use above, or is the causality the reverse?

File: src/sinks/honeycomb.rs [hunk 6]
```diff
@@ -18,10 +17,11 @@ use crate::{
     },
 };
 
-static HOST: Lazy<Uri> = Lazy::new(|| Uri::from_static("https://api.honeycomb.io/1/batch"));
-
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub(super) struct HoneycombConfig {
+    #[serde(skip, default = "default_endpoint")]
+    endpoint: String,
+
```
Comment (bruceg) [inline src/sinks/honeycomb.rs:24]: Ditto needs documentation.

File: src/sinks/util/request_builder.rs [hunk 7]
```diff
@@ -67,6 +67,10 @@ pub trait RequestBuilder<Input> {
         &self,
         events: Self::Events,
     ) -> Result<EncodeResult<Self::Payload>, Self::Error> {
+        // TODO: Should we add enough bounds on `Self::Events` that we could automatically derive event count/event byte
+        // size, and then we could generate `BatchRequestMetadata` and pass it directly to `build_request`? That would
+        // obviate needing to wrap `payload` in `EncodeResult`, although practically speaking.. the name would be kind
+        // of clash-y with `Self::Metadata`.
```
Comment (bruceg) [inline src/sinks/util/request_builder.rs:73]: I like the idea of automatically generating the data. I think we could handle the name clash if the results are worth it. In any case, noting this TODO needs a follow up.

File: src/sinks/util/udp.rs [hunk 8]
```diff
@@ -223,6 +226,19 @@ impl tower::Service<BytesMut> for UdpService {
             // TODO: Add reconnect support as TCP/Unix?
             let result = udp_send(&mut socket, &msg).await.context(SendSnafu);
             let _ = sender.send(socket);
+
+            if result.is_ok() {
+                // TODO: This is obviously not happening before things like compression, etc, so it's currently a
```
Comment (bruceg) [inline src/sinks/util/udp.rs:231]: Not so much a TODO here as a note or warning.

**Review by blt — APPROVED** [review]

**Review by bruceg — COMMENTED** [review]
The AWS `ClientBuilder` changes look like a big win, though I would have preferred seeing that broken out to make this easier to review. This looks like a good improvement, though I have a number of questions…

File: src/sinks/splunk_hec/logs/integration_tests.rs [hunk 9]
```diff
@@ -127,11 +129,9 @@ async fn splunk_insert_message() {
 
     let message = random_string(100);
     let (batch, mut receiver) = BatchNotifier::new_with_receiver();
-    let event = LogEvent::from(message.clone())
-        .with_batch_notifier(&batch)
-        .into();
+    let event = Event::from(message.clone()).with_batch_notifier(&batch);
```
[the remark this answers is not in the document]
Comment (tobz) [reply src/sinks/splunk_hec/logs/integration_tests.rs:132]: I guess I'm not sure what you mean. Always using `.into()`, or doing it implicitly via generic parameters on functions such as `T: Into<Event>`? Instead of using `Event::from(...)`?

File: src/sinks/util/metadata.rs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+#[derive(Clone, Debug)]
+pub struct BatchRequestMetadata {
+    /// Number of events represented by this batch request.
+    pub event_count: usize,
+    /// Size, in bytes, of the in-memory representation of all events in this batch request.
+    pub event_byte_size: usize,
+    /// Uncompressed size, in bytes, of the encoded events in this batch request.
+    pub encoded_uncompressed_size: usize,
+    /// Compressed size, in bytes, of the encoded events in this batch request, if compression was performed.
+    pub encoded_compressed_size: Option<usize>,
```
[the remark this answers is not in the document]
Comment (tobz) [reply src/sinks/util/metadata.rs:13]: Yeah. I was pigeonholing the data type on the name itself, trying to indicate that compression isn't always certain and thus the optionality of it. Renaming it to `encoded_wire_size`, though, would provide the same data and avoid the ambiguity.

Getting the wire size -- compressed or not -- is gonna be something I need to do for the raw network metrics, so I'll make that change here now.

[the remark this answers is not in the document]
Comment (tobz) [reply src/sinks/util/metadata.rs:13]: Hmmm, perhaps? I have an idea of how to try and make this a little more builder-y, to avoid the chance for incorrect values to get stored... so let me give that a whirl. 👍🏻

File: src/sinks/splunk_hec/logs/integration_tests.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (bruceg) [reply src/sinks/splunk_hec/logs/integration_tests.rs:132]: What I mean is that an `Event` is a generic data type. If you want a log, create a `LogEvent`, etc. From what I understand, `Event::from(String)` is a relic from when there weren't other kinds of events.

File: src/sinks/util/metadata.rs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+// TODO: Make this struct the object which emits the actual internal telemetry i.e. events sent, bytes sent, etc.
```
[the remark this answers is not in the document]
Comment (tobz) [reply src/sinks/util/metadata.rs:16]: My thought is semi-yes.

I'm kind of torn: I think having a type that holds all the metric values be the one to actually emit them is a better separation of concerns than encoding it into `DriverResponse`... but it would, at this point, essentially be a no-op in functionality a lot of refactoring that just inverts from `DriverResponse` to `BatchRequestMetadata`.

This is all further complicated by the notion that, at some point, we should be tracking all bytes sent from a sink -- including retries -- and `DriverResponse` isn't quite the right level to do that, given that it only comes when the request is successful and has no access to retries, etc.

Using `BatchRequestMetadata`, or some additional type that can be passed around as a drop guard, as part of the request itself, would be a natural pathway, and if we're injecting things at that level, having them duplicated at the end with the `DriverResponse` machinery feels like a lot of conceptual overlap.

Long story short, I'd like to have `BatchRequestMetadata` -- probably simply renamed `RequestMetadata` -- be a successor to `DriverResponse` and get extended further. I'm not sure I should explore that design in this PR, though. Hence the TODO.

File: src/sinks/util/udp.rs (hunk 8 above)
Comment (tobz) [reply src/sinks/util/udp.rs:231]: Fair point.

File: src/sinks/splunk_hec/logs/integration_tests.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (tobz) [reply src/sinks/splunk_hec/logs/integration_tests.rs:132]: Ah, yeah, totally.

The reason I did it like this is that `run_and_assert_sink_compliance` takes a `Stream<Item = I> where I: Into<EventArray>`, and there is no conversion to go directly from `LogEvent` to `EventArray` but there was one for `Event`. I can go ahead and add one for `LogEvent` and `Metric`, though, now that I'm thinking about it.

[the remark this answers is not in the document]
Comment (bruceg) [reply src/sinks/splunk_hec/logs/integration_tests.rs:132]: Ah, hmm, and doing `LogEvent::from(…).into()` doesn't work because of the double `Into` at work there. The alternate is `Event::Log(LogEvent::from(X))` or `Event::from(LogEvent)` but that could get wordy.

File: src/sinks/gcs_common/service.rs (hunk 5 above)
Comment (tobz) [reply src/sinks/gcs_common/service.rs:48]: Factored out the custom type back to being just a tuple, which matches the other implementations of `RequestBuilder`. It was also going to look weird since the request builder was setting `GcsMetadata` as a field on `GcsRequest`.. and then I was going to be setting another field as a `BatchRequestMetadata`.

Just felt weird/repetitive.

File: src/sinks/datadog_archives.rs (hunk 1 above)
Comment (tobz) [reply src/sinks/datadog_archives.rs:589]: My comment here explains the rationale: https://github.com/vectordotdev/vector/pull/12755#discussion_r881134240

File: src/sinks/gcp/cloud_storage.rs (hunk 2 above)
Comment (tobz) [reply src/sinks/gcp/cloud_storage.rs:223]: Covered by https://github.com/vectordotdev/vector/pull/12755#discussion_r881134240.

File: src/sinks/gcp/stackdriver_logs.rs (hunk 3 above)
Comment (tobz) [reply src/sinks/gcp/stackdriver_logs.rs:41]: It's marked `skip` and always has a default, so it's not part of the user-facing configuration.

File: src/sinks/honeycomb.rs (hunk 6 above)
Comment (tobz) [reply src/sinks/honeycomb.rs:24]: It's marked `skip` and always has a default, so it's not part of the user-facing configuration.

File: src/sinks/util/metadata.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (tobz) [reply src/sinks/util/metadata.rs]: I've reworked `RequestMetadata` (neé `BatchRequestMetadata`) a bit and included this idea in it, as part of 6b8a34b9f.

[the remark this answers is not in the document]
Comment (tobz) [reply src/sinks/util/metadata.rs]: I'e significantly reworked `RequestMetadata` in 6b8a34b9f such that it must be built from something that is `EventCount + ByteSizeOf` to get the event count/event byte size, and then either given an `&EncodeResult` or a single `NonZeroUsize` for both encoded/wire size.

Now there is no way to pass in raw `usize` values that could be incorrectly set to zero, although it doesn't stop the source of those values from being zero... although those values could propagate to other places and cause issues, so I'm not sure `RequestMetadata` is the place to make a stand against zero values in general.

File: src/aws/mod.rs [hunk 12]
```diff
@@ -144,3 +151,159 @@ impl AsyncSleep for TokioSleep {
...
+        // sends the request out over the wire. We'll read the shared atomic counter, which will contain the number of
+        // bytes "read", aka the bytes it actually sent, if and only if we get back a successful response.
+        let maybe_bytes_sent = if self.enabled {
+            let (callback, shared_bytes_sent) = BodyCaptureCallback::new();
+            req.http_mut().body_mut().with_callback(Box::new(callback));
+
+            Some(shared_bytes_sent)
+        } else {
+            None
+        };
```
[the remark this answers is not in the document]
Comment (tobz) [reply src/aws/mod.rs]: Addressed in 6b8a34b9f.

File: src/aws/mod.rs [hunk 13]
```diff
@@ -144,3 +151,159 @@ impl AsyncSleep for TokioSleep {
...
+    }
+}
+
+impl Drop for BodyCaptureCallback {
+    fn drop(&mut self) {
+        // This is where we actually emit. We specifically emit here, and not in `trailers`, because despite the
+        // documentation that `trailers` is called after all chunks of the body are successfully read, `hyper` won't
+        // continue polling a body if it knows it's gotten all the available bytes i.e. it doesn't necessarily drive it
+        // until `poll_data` returns `None`. This means the only consistent place to know that the body is "done" is
+        // when it's dropped.
```
[the remark this answers is not in the document]
Comment (tobz) [reply src/aws/mod.rs:290]: Yep. Chatted with one of the engineers working on the AWS Rust SDK who said they'd be reworking this area of the code in a way that would allow the request body to be wrapped with a custom type on our end, so as to avoid needing the `BodyCallback` stuff at all, given that it can't function correctly as implemented, given what `hyper` does.

File: src/http.rs [hunk 14]
```diff
@@ -256,6 +256,28 @@ impl Auth {
...
+            // `http::Uri` ensures that we always get "http" or "https" if the URI is created with a well-formed scheme,
+            // but it also supports arbitrary schemes, which is where we bomb out down here, since we can't generate a
+            // static string for an arbitrary input string... and anything other than "http" and "https" makes no sense
+            // for an HTTP client anyways.
+            s => panic!("invalid URI scheme for HTTP client: {}", s),
+        },
+        // If there's no scheme, we just use "http" since it provides the most semantic relevance without inadvertantly
+        // implying things it can't know i.e. returning "https" when we're not actually sure HTTPS was used.
+        None => "http",
+    }
```
[the remark this answers is not in the document]
Comment (tobz) [reply src/http.rs]: Addressed in 6b8a34b9f.

File: src/sinks/util/udp.rs (hunk 8 above)
Comment (tobz) [reply src/sinks/util/udp.rs]: Addressed in 6b8a34b9f.

File: src/internal_events/aws.rs [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    pub endpoint: &'a str,
+}
+
+impl<'a> InternalEvent for AwsBytesSent<'a> {
+    fn emit(self) {
+        let region = self
+            .region
+            .as_ref()
+            .map(|r| r.as_ref().to_string())
+            .unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (tobz) [reply src/internal_events/aws.rs]: Addressed in 6b8a34b9f.

File: src/sinks/splunk_hec/logs/integration_tests.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (tobz) [reply src/sinks/splunk_hec/logs/integration_tests.rs]: Yeah. :/

I ended up addressing this in 6b8a34b9f by simply adding `From` conversions for `LogEvent` (and siblings) into `EventArray`, via `Event::from`.

File: src/sinks/util/request_builder.rs (hunk 7 above)
Comment (tobz) [reply src/sinks/util/request_builder.rs:73]: Yeah, to say it loud so it's not just a thought in my brain: I think `RequestMetadata` could actually take on a generic parameter for a `data` field, _and_ a field for finalizers.

Right now, `RequestBuilder` bifurcates the inputs into metadata and events, and then encodes the events, and passes the encoded events and the bifurcated metadata to the method that inevitably builds the request.

Instead, we could make it so that the metadata/event splitting step stays as-is, but the middle step generates the count/size stuff, wraps the user-provided metadata, and then passes the wrapped metadata (`RequestMetadata<Self::Metadata>`) and the payload at the very end.

As far as the finalizers go, we basically always factor them out of the input events and pass them within `Self::Metadata`, so we might as well just do that for the implementor as well. This would also let us implement `Finalizable` for `RequestMetadata<T>`, removing yet another needlessly-broken-out field on the "true" request value i.e. `GcsRequest`.

File: lib/vector-buffers/src/lib.rs [hunk 16]
```diff
@@ -84,6 +84,18 @@ pub trait EventCount {
...
+    fn event_count(&self) -> usize {
+        self.len()
+    }
+}
+
+impl<T> EventCount for &Vec<T> {
+    fn event_count(&self) -> usize {
+        self.len()
+    }
+}
```
Comment (bruceg) [inline lib/vector-buffers/src/lib.rs:97]: Could this be used with `Vec<EventArray>` to do the wrong thing?

Comment (tobz) [reply lib/vector-buffers/src/lib.rs:97]: It could, yes. I'll file an issue to re-examine how I approached wiring up `EventCount` to accomplish my work in `RequestMetadata.

I was really struggling to find the right fit in terms of getting the correct count of events for arbitrary event inputs to `RequestBuilder`. In particular, I did this because the `loki` sink -- maybe others? I was just changing/adding trait impls until `cargo check` was happy -- uses `Vec<LokiRecord>` for its input events to the request builder.

In hindsight, the simpler solution was probably to bound the `EventCount` impl for `T: EventCount` and do a sum, and then impl `EventCount` for `LokiRecord`. I'll probably anchor the issue around taking that approach, unless other usages require a different/cleaner approach.

File: src/sinks/gcp/stackdriver_metrics.rs (hunk 4 above)
Comment (tobz) [reply src/sinks/gcp/stackdriver_metrics.rs:3]: Opened #12856 to track this.

File: src/sinks/util/request_builder.rs (hunk 7 above)
Comment (tobz) [reply src/sinks/util/request_builder.rs:73]: Opened #12857 to track this.
